slideflow/io/preservedsite/crossfolds.py

```diff
@@ -2,12 +2,129 @@
+        list_of_all_combos = list()
+        for i in range(most_possible_sites_in_one_fold):
+            list_of_all_combos += list(combinations(unique_sites, i+1))
+        list_of_proper_combos = list(combinations(list_of_all_combos, crossfolds))
+        removal_list = list()
+        for item in list_of_proper_combos:
+            item_length = 0
+            list_of_items_in_a_combo = list()
+            for item2 in item:
+                item_length += len(item2)
+                list_of_items_in_a_combo.extend(item2)
+            if item_length < len(unique_sites) or item_length > len(unique_sites):
+                removal_list.append(item)
+            if len(list_of_items_in_a_combo) != len(set(list_of_items_in_a_combo)):
+                removal_list.append(item)
+        for item in set(removal_list):
+            list_of_proper_combos.remove(item)
+    
+    # split of values per site
+    data_dict = dict()
+    for site in unique_sites:
+        dict_of_values = dict()
+        sum = 0
+        for value in values:
+            dict_of_values[value] = len(newData[((newData[site_column] == site) & (newData[category] == value))])
+            sum += len(newData[((newData[site_column] == site) & (newData[category] == value))])
+        dict_of_values['total'] = sum
+        data_dict[site] = dict_of_values
+
+    # error associated to each possible combo
+    per_fold_size_target_ratio = float(1)/crossfolds
+    per_site_target_ratio = float(1)/len(values)
+    per_combo_errors = dict()
+    dictionary_of_split = dict()
+    for combo in list_of_proper_combos:
+        dictionary_of_split[combo] = dict()
+        mean_square_error = 0
+        count = 0
+        for fold in combo:
+            sum2 = 0
+            dictionary_of_split[combo][fold] = {value: 0 for value in values}
+            dictionary_of_split[combo][fold]['total'] = 0
+            for site in fold:
+                for key in data_dict[str(site)].keys():
+                    dictionary_of_split[combo][fold][key] += data_dict[site][key]
+            sum2 += dictionary_of_split[combo][fold]['total']
+            for k in dictionary_of_split[combo][fold].keys():
+                 dictionary_of_split[combo][fold][k] = dictionary_of_split[combo][fold][k]/float(dictionary_of_split[combo][fold]['total'])
+                 fold_site_value = dictionary_of_split[combo][fold][k]
+                 count += 1
+                 mean_square_error += (fold_site_value-per_site_target_ratio)**2
+        for fold in combo:
+            fold_total = dictionary_of_split[combo][fold]['total']/(float(sum2))
+            mean_square_error += (fold_total - per_fold_size_target_ratio)**2
+            count += 1
+        mean_square_error = mean_square_error/count
+        per_combo_errors[combo] = mean_square_error
+    
+    # isolate best combo by error
+    min = 100000000000000000000000000000000000000000000

```

Comment:
I see. In that case, you should remove the `min` variable (having a manually set large value is clunky). Instead, you can just sort the dictionary and find the lowest value. So instead of:

``` python
min = 100000000000000000000000000000000000000000000
best_combo = None
for key, value in per_combo_errors.items():
    if value < min:
        best_combo = key
        min = value
    else:
        pass
```

do

``` python
best_combo = min(per_combo_errors, key=per_combo_errors.get)
```